src/net.cpp

```diff
@@ -906,36 +906,50 @@ std::pair<size_t, bool> CConnman::SocketSendData(CNode& node) const
 {
     auto it = node.vSendMsg.begin();
     size_t nSentSize = 0;
-
-    while (it != node.vSendMsg.end()) {
-        const auto& data = *it;
-        assert(data.size() > node.nSendOffset);
+    bool have_unsent_data{false};
+
+    while (true) {
+        if (it != node.vSendMsg.end()) {
+            // If possible, move one message from the send queue to the transport. This fails when
+            // there is an existing message still being sent.
+            size_t memusage = it->GetMemoryUsage();
+            if (node.m_transport->SetMessageToSend(*it)) {
+                // Update memory usage of send buffer (as *it will be deleted).
+                node.m_send_memusage -= memusage;
+                ++it;
+            }
```

Comment:
8d1a91f6394f00b880613268a9f2f3164ef914e2: IIUC even though `!node.vSendMsg.empty()`, we don't want `else { have_unsent_data = true }` here because:

1. `SetMessageToSend` _may_ return `false` for other reasons (though afaik it can't at the moment)
2. If it does return `false`, then `GetBytesToSend` below populates `data`, tries to send it and then updates `have_unsent_data` appropriately.